### PR TITLE
Add POST support to /forecasts

### PR DESCRIPTION
**Pull request by acron0**

### CODE REVIEW

File: src/witan/app/forecast.clj [hunk 1]
```diff
@@ -16,15 +17,16 @@
...
-             :created (util/java-Date-to-ISO-Date-Time created))))
+  (let [cleaned (-> forecast
+                    (dissoc :descendant_id
+                            :in_progress
+                            :series_id
+                            :created)
+                    (assoc :descendant-id descendant_id
+                           :in-progress? in_progress
+                           :series-id series_id
+                           :created (util/java-Date-to-ISO-Date-Time created)))]
```
Comment (otfrom) [inline src/witan/app/forecast.clj]: Why the change from snake_case to kebab-case?

Comment (acron0) [inline src/witan/app/forecast.clj]: We started with kebab-case as more idiomatic, Cassandra doesn't appear to support it so, so schema is snake_case and have this step to coerce everything back.

File: src/witan/app/forecast.clj [hunk 2]
```diff
@@ -103,9 +105,33 @@
...
 (defresource forecasts
+  util/json-resource
   :allowed-methods #{:get :post}
-  :available-media-types ["application/json"]
-  :handle-ok (fn [_] (s/validate
-                      [ws/Forecast]
-                      (map ->Forecast (get-forecasts)))))
+  :processable? (util/post!-processable-validation ws/NewForecast)
+  :exists? (fn [ctx]
+             (if (= :post (-> ctx :request :request-method))
```
Comment (otfrom) [inline src/witan/app/forecast.clj]: I bet `(= :post (-> ctx :request :request-method)` will want to be a little helper function somewhere `http-post?`

File: src/witan/app/forecast.clj [hunk 3]
```diff
@@ -103,9 +105,33 @@
...
+               (let [{:keys [name]} (-> ctx :request :body-params)
```
Comment (otfrom) [inline src/witan/app/forecast.clj]: and `(-> ctx :request :body-params)` as a helper too

File: src/witan/app/forecast.clj [hunk 4]
```diff
@@ -103,9 +105,33 @@
...
+                     owner (-> ctx :request :identity)]
+                 (not-empty (c/exec (find-forecast-by-name-and-owner name owner))))
+               true))
+  ;;
+  :if-match-exists? false
+  :if-unmodified-since-exists? false
+  :if-none-match-exists? (fn [c] (= :post (-> c :request :request-method)))
```
Comment (otfrom) [inline src/witan/app/forecast.clj]: `ctx` rather than `c` for consistency?

File: src/witan/app/handler.clj [hunk 5]
```diff
@@ -44,7 +44,10 @@
 (defn add-token!
   [user-id tokens]
   (let [token (user/random-token)
-        ttl (t/days 28)]
+        ttl (t/days 28)
+        existing (some #(if (= user-id (-> % second :user)) %) @tokens)]
+    (when existing ;; remove an existing token for this user
+      (swap! tokens dissoc (first existing)))
```
Comment (otfrom) [inline src/witan/app/handler.clj]: is this managing state across the system? Should this be stored in a session store of some sort?

File: src/witan/app/schema.clj [hunk 6]
```diff
@@ -114,6 +114,13 @@
    (s/required-key :forecast-id) IdType
    (s/required-key :forecast-version) s/Int})
 
+(def NewForecast
+  "Schema for creating a new forecast"
+  {(s/required-key :name)              s/Str
+   (s/required-key :model-id)         (s/either IdType s/Str)
+   (s/optional-key :description)       s/Str
+   (s/optional-key :model-properties) [ModelPropertyValue]})
```
Comment (otfrom) [inline src/witan/app/schema.clj]: it feels to me like you've done a fair bit of work here that won't agree w/ align-clj-let or cljfmt. Is that the case?

File: src/witan/app/util.clj [hunk 7]
```diff
@@ -2,7 +2,34 @@
...
+            [clj-time.coerce   :as tc]
+            [schema.core :as s]))
+
+(defn java-Date-to-ISO-Date-Time
+  "Converts a java.util.Date to an schema-contrib/ISO-Date-Time"
+  [datetime]
+  (tf/unparse (tf/formatters :date-hour-minute-second)
+              (tc/from-date datetime)))
+
+(defn is-json?
```
Comment (otfrom) [inline src/witan/app/util.clj]: `json?`

File: src/witan/app/util.clj [hunk 8]
```diff
@@ -2,7 +2,34 @@
...
+  [string]
+  (try
+    (do (json/read-str str) true)
+    (catch Exception e false)))
+
+(defn post!-processable-validation
+  "In the context of a POST, checks the body params against a schema"
+  [schema]
+  (fn [ctx]
+    (if (= :post (-> ctx :request :request-method))
```
Comment (otfrom) [inline src/witan/app/util.clj]: our `post?` helper again

File: src/witan/app/util.clj [hunk 9]
```diff
@@ -2,7 +2,34 @@
...
+      (let [params (-> ctx :request :body-params)]
```
Comment (otfrom) [inline src/witan/app/util.clj]: our body params helper again

File: src/witan/app/util.clj [hunk 10]
```diff
@@ -10,10 +37,25 @@
...
-              (tc/from-date datetime)))
+;;;;;;;;;;;;;;;;
+
+;; Be aware of [this issue](https://github.com/clojure-liberator/liberator/issues/94)
+;; when working on the 'malformed' parts
+
+(def json-resource
+  {:available-media-types ["application/json"]
+   :known-content-type? (fn [ctx]
+                          (if (= :post (-> ctx :request :request-method))
```
Comment (otfrom) [inline src/witan/app/util.clj]: `post?`

File: src/witan/app/util.clj [hunk 11]
```diff
@@ -10,10 +37,25 @@
...
+                            (= "application/json" (-> ctx :request :content-type)))
```
Comment (otfrom) [inline src/witan/app/util.clj]: `json-content?`

File: src/witan/app/forecast.clj [hunk 12]
```diff
@@ -103,9 +105,33 @@
...
-  :handle-ok (fn [_] (s/validate
-                      [ws/Forecast]
-                      (map ->Forecast (get-forecasts)))))
+  :processable? (util/post!-processable-validation ws/NewForecast)
+  :exists? (fn [ctx]
+             (if (= :post (-> ctx :request :request-method))
+               (let [{:keys [name]} (-> ctx :request :body-params)
+                     owner (-> ctx :request :identity)]
+                 (not-empty (c/exec (find-forecast-by-name-and-owner name owner))))
+               true))
```
Comment (otfrom) [inline src/witan/app/forecast.clj]: if you are hitting the database in exists (usually for a get) I'd put the results of that query into the context as soon as possible so that downstream functions only need to check the context and not the database.

File: src/witan/app/handler.clj (hunk 5 above)
Comment (acron0) [inline src/witan/app/handler.clj]: So this is just the authn tokens. @elisehuard and I spoke about this and reasoned that for the MVP we could "get away with" leaving them here, but as soon as we do any kind of load balancing this will die, so we need an idea of how to scale this solution at some point.

File: src/witan/app/schema.clj (hunk 6 above)
Comment (acron0) [inline src/witan/app/schema.clj]: cljfmt doesn't seem to complain :|

File: src/witan/app/handler.clj (hunk 5 above)
Comment (otfrom) [inline src/witan/app/handler.clj]: yeah, this will become a pain quickly. There is an example of using a C* session store here:

https://github.com/MastodonC/kixi.hecuba/blob/master/src/clj/kixi/hecuba/session.clj

Comment (acron0) [inline src/witan/app/handler.clj]: Card is created: https://trello.com/c/6xTPikYC

File: src/witan/app/forecast.clj (hunk 1 above)
Comment (otfrom) [inline src/witan/app/forecast.clj]: hmm... that would be why everything is snake_case in hecuba. Saves trouble in the json as well IIRC

**Comment by otfrom** [comment]
:shipit:

**Comment by otfrom** [comment]
I AM THE GOD OF THE BIKESHED! :-P
